src/expr/parser.rs

```diff
@@ -25,34 +47,46 @@ named!(i64_literal<i64>,
   )
 );
 
-named!(factor<i64>,
+named!(factor<Expr>,
   alt!(
-    i64_literal
+    expr_literal
   | parens
   )
 );
 
 // we define acc as mutable to update its value whenever a new term is found
-named!(term <i64>,
+named!(term <Expr>,
   chain!(
     mut acc: factor  ~
              many0!(
                alt!(
-                 tap!(mul: preceded!(tag!("*"), factor) => acc = acc * mul) |
-                 tap!(div: preceded!(tag!("/"), factor) => acc = acc / div)
+                 map!(preceded!(tag!("*"), factor), |mul| {
+                                                            let acc_expr = mem::replace(&mut acc, Expr::Literal(0));
+                                                            acc = Expr::BinOp(Box::new(acc_expr), BinOp::Times, Box::new(mul))
+                                                          })|
```

Comment:
You use this pattern a lot -- consider abstracting it out as some sort of "grow expression".